discovery_server.go

```diff
@@ -0,0 +1,314 @@
+			d.outputError("command_error", fmt.Sprintf("First command must be HELLO, but got '%s'", cmd))
+			continue
+		}
+
+		switch cmd {
+		case "HELLO":
+			d.hello(fullCmd[6:])
+		case "START":
+			d.start()
+		case "LIST":
+			d.list()
+		case "START_SYNC":
+			d.startSync()
+		case "STOP":
+			d.stop()
+		case "QUIT":
+			d.impl.Quit()
+			d.outputOk("quit")
+			return nil
+		default:
+			d.outputError("command_error", fmt.Sprintf("Command %s not supported", cmd))
+		}
+	}
+}
+
+func (d *DiscoveryServer) hello(cmd string) {
+	if d.initialized {
+		d.outputError("hello", "HELLO already called")
+		return
+	}
+	re := regexp.MustCompile(`(\d+) "([^"]+)"`)
+	matches := re.FindStringSubmatch(cmd)
+	if len(matches) != 3 {
+		d.outputError("hello", "Invalid HELLO command")
+		return
+	}
+	d.userAgent = matches[2]
+	if v, err := strconv.ParseInt(matches[1], 10, 64); err != nil {
+		d.outputError("hello", "Invalid protocol version: "+matches[2])
+		return
+	} else {
+		d.reqProtocolVersion = int(v)
+	}
+	if err := d.impl.Hello(d.userAgent, 1); err != nil {
+		d.outputError("hello", err.Error())
+		return
+	}
+	d.output(&genericMessageJSON{
+		EventType:       "hello",
+		ProtocolVersion: 1, // Protocol version 1 is the only supported for now...
+		Message:         "OK",
+	})
+	d.initialized = true
+}
+
+func (d *DiscoveryServer) start() {
+	if d.started {
+		d.outputError("start", "Discovery already STARTed")
+		return
+	}
+	if d.syncStarted {
+		d.outputError("start", "Discovery already START_SYNCed, cannot START")
+		return
+	}
+	if err := d.impl.Start(); err != nil {
+		d.outputError("start", "Cannot START: "+err.Error())
+		return
+	}
+	d.started = true
+	d.outputOk("start")
+}
+
+func (d *DiscoveryServer) list() {
+	if !d.started {
+		d.outputError("list", "Discovery not STARTed")
+		return
+	}
+	if d.syncStarted {
+		d.outputError("list", "discovery already START_SYNCed, LIST not allowed")
+		return
+	}
+	if ports, err := d.impl.List(); err != nil {
+		d.outputError("list", "LIST error: "+err.Error())
+		return
+	} else {
+		type listOutputJSON struct {
+			EventType string  `json:"eventType"`
+			Ports     []*Port `json:"ports"`
+		}
+		d.output(&listOutputJSON{
+			EventType: "list",
+			Ports:     ports,
+		})
+	}
+}
+
+func (d *DiscoveryServer) startSync() {
+	if d.syncStarted {
+		d.outputError("start_sync", "Discovery already START_SYNCed")
+		return
+	}
+	if d.started {
+		d.outputError("start_sync", "Discovery already STARTed, cannot START_SYNC")
+		return
+	}
+	d.syncChannel = make(chan interface{}, 10) // buffer up to 10 events
+	if err := d.impl.StartSync(d.syncEvent); err != nil {
+		d.outputError("start_sync", "Cannot START_SYNC: "+err.Error())
+		close(d.syncChannel) // do not leak channel...
+		d.syncChannel = nil
+		return
+	}
+	d.syncStarted = true
+	d.outputOk("start_sync")
+	go d.consumeEvents(d.syncChannel)
+}
+
+func (d *DiscoveryServer) consumeEvents(c <-chan interface{}) {
+	for e := range c {
+		d.output(e)
+	}
+}
```

Comment:
I'd avoid a separate function for this, technically no other command must start consuming events so it should be safer to just run it like so:

```
go func() {
	for e := range c {
		d.output(e)
	}
}()
```

This way we avoid calling it somewhere else in future modifications.